`AnyHolderSet` and `NotHolderSet` cannot be datagenned

**Minecraft Version:** 1.21

**NeoForge Version:** 21.4.51-beta

**Exception:**

```
Caused by: java.lang.ClassCastException: class net.minecraft.core.RegistrySetBuilder$UniversalLookup cannot be cast to class net.minecraft.core.HolderLookup$RegistryLookup (net.minecraft.core.RegistrySetBuilder$UniversalLookup and net.minecraft.core.HolderLookup$RegistryLookup are in module minecraft@1.21.1 of loader 'TRANSFORMER' @989da1)
```

**Description of issue:**

Attempting to use either `AnyHolderSet` or `NotHolderSet` in datagen leads to a crash. The reason for this is both require a `HolderLookup.RegistryLookup` to construct, and the Neo provided method to obtain a `RegistryLookup` does an unchecked cast from a `HolderGetter` to a `HolderLookup.RegistryLookup` which fails during datagen (I've not verified if that cast ever works at runtime,  but I assume it must otherwise it wouldn't be there).

One commonly attempted fix is simply passing in `null` to the constructor since we don't *need* the `RegistryLookup` information during datagen, but this leads to a crash in Mojang's `canSerializeIn` which attempts to use the lookup during datagen.

**Workarounds:**

* The most general workaround is to construct a dummy `RegistryLookup` by either wrapping the `HolderLookup` or throwing in most methods, an example is here: https://github.com/SlimeKnights/TinkersConstruct/blob/e335dfb1c81b7698fe534b1b74ddf4ed85692bed/src/main/java/slimeknights/tconstruct/world/data/WorldgenProvider.java#L372
* A simpler workaround is extending `NotHolderSet`/`AnyHolderSet` for a datagennable version by making `canSerializeIn` always return `true`, thus making the constructor able to accept `null` during datagen, example here: https://github.com/SlimeKnights/TinkersConstruct/blob/85478dc6122864c6ac8d9001cc025882bb506714/src/main/java/slimeknights/tconstruct/world/data/WorldgenProvider.java#L363

**Related Issues:**

* Original Forge issue, closed without a fix: MinecraftForge/MinecraftForge#9634
* PR fixing the Forge issue, uses an approach different from either of my workarounds that is worth considering: MinecraftForge/MinecraftForge#9663

--- a/patches/net/minecraft/core/RegistrySetBuilder.java.patch
+++ b/patches/net/minecraft/core/RegistrySetBuilder.java.patch
@@ -52,15 +52,42 @@
              p_256495_.forEach(p_256603_ -> builder.put(p_256603_.location(), registrysetbuilder$universallookup));
              return new RegistrySetBuilder.BuildState(
                  registrysetbuilder$universalowner, registrysetbuilder$universallookup, builder.build(), new HashMap<>(), list
-@@ -259,6 +_,11 @@
-                 @Override
+@@ -260,6 +_,11 @@
                  public <S> HolderGetter<S> lookup(ResourceKey<? extends Registry<? extends S>> p_255961_) {
                      return (HolderGetter<S>)BuildState.this.registries.getOrDefault(p_255961_.location(), BuildState.this.lookup);
-+                }
+                 }
 +
 +                @Override
 +                public <S> Optional<HolderLookup.RegistryLookup<S>> registryLookup(ResourceKey<? extends Registry<? extends S>> registry) {
 +                    return Optional.ofNullable((HolderLookup.RegistryLookup<S>) BuildState.this.registries.get(registry.location()));
-                 }
++                }
              };
          }
+ 
+@@ -417,6 +_,26 @@
+         @Override
+         public Optional<Holder.Reference<Object>> get(ResourceKey<Object> p_256303_) {
+             return Optional.of(this.getOrCreate(p_256303_));
++        }
++
++        @Override
++        public Stream<Holder.Reference<Object>> listElements() {
++            return holders.values().stream();
++        }
++
++        @Override
++        public Stream<ResourceKey<Object>> listElementIds() {
++            return holders.keySet().stream();
++        }
++
++        @Override
++        public Stream<HolderSet.Named<Object>> listTags() {
++            return Stream.empty();
++        }
++
++        @Override
++        public Stream<TagKey<Object>> listTagIds() {
++            return Stream.empty();
+         }
+ 
+         <T> Holder.Reference<T> getOrCreate(ResourceKey<T> p_256298_) {

--- a/src/main/java/net/neoforged/neoforge/registries/holdersets/AnyHolderSet.java
+++ b/src/main/java/net/neoforged/neoforge/registries/holdersets/AnyHolderSet.java
@@ -86,7 +86,7 @@ public record AnyHolderSet<T>(HolderLookup.RegistryLookup<T> registryLookup) imp
 
     @Override
     public boolean canSerializeIn(HolderOwner<T> holderOwner) {
-        return this.registryLookup.canSerializeIn(holderOwner);
+        return true;
     }
 
     @Override

--- a/src/main/java/net/neoforged/neoforge/registries/holdersets/NotHolderSet.java
+++ b/src/main/java/net/neoforged/neoforge/registries/holdersets/NotHolderSet.java
@@ -115,7 +115,7 @@ public class NotHolderSet<T> implements ICustomHolderSet<T> {
 
     @Override
     public boolean canSerializeIn(HolderOwner<T> holderOwner) {
-        return this.registryLookup.canSerializeIn(holderOwner);
+        return this.value.canSerializeIn(holderOwner);
     }
 
     @Override

--- a/src/main/resources/META-INF/injected-interfaces.json
+++ b/src/main/resources/META-INF/injected-interfaces.json
@@ -74,6 +74,9 @@
   "net/minecraft/core/Registry": [
     "net/neoforged/neoforge/registries/IRegistryExtension<T>"
   ],
+  "net/minecraft/core/RegistrySetBuilder$UniversalLookup": [
+    "net/minecraft/core/HolderLookup<java.lang.Object>"
+  ],
   "net/minecraft/core/component/DataComponentHolder": [
     "net/neoforged/neoforge/common/extensions/IDataComponentHolderExtension"
   ],

